### ISSUE TITLE
Controlled Tables are sluggish due to excessive rendering

### ISSUE THREAD

**Issue by VictorGlindasPaf** [adobe/react-spectrum#5707]
### Provide a general summary of the issue here

Selecting rows in a controlled `Table` causes the entire table to render multiple times, causing performance issues in anything but the simplest examples.

I'm also not sure why the rows/cells are rerendered on hover? This is also not super great for performance, but is less of a concern since it's local to a single row.

### 🤔 Expected Behavior?

The table should only rerender the affected rows. Preferably once.

### 😯 Current Behavior

It rerenders the entire table multiple times.

### 🖥️ Steps to Reproduce

[CodeSandbox Reproduction](https://codesandbox.io/p/sandbox/hardcore-curie-cvj8kj?file=%2Fsrc%2FApp.js%3A34%2C3-42%2C6)

Open the console and increase the number of rows  to see the render performance logging.

### Version

1.0.1

### What browsers are you seeing the problem on?

Chrome

### What operating system are you using?

MacOS Sonoma 14.0

**Comment by steveoh** [comment]
This is affecting our implementation also. Is there a recommended row size for this table implementation or advice for working around the performance problem?

**Comment by steveoh** [comment]
I notice the lag in a production build with 550 records and 12 columns with multiple selections enabled.

We'll have to check out the virtualization but may wait for documentation.

**Comment by henkkasoft** [comment]
> Otherwise, when you have a lot of items, it may be a good idea to virtualize the table. We have some tools for doing that now. They are somewhat new, so they aren't in the docs yet, but you can check it out in the source code https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/s2/src/TableView.tsx#L329

Do you know when this virtualizing would be documented? And is this available in react-aria-components?

We have implemented button to open rows and the usability of opening the row gets pretty bad when row count is getting bigger. And it seems to render twice the table after the open row button is pressed.
Production build testing opening a single row in Table takes approx:
200 rows table = less than 1 sec
500 rows table = 2 sec
1000 rows table = 5 sec
1500 rows table = over 10 sec

So the usage is not smooth even with a couple of hundred rows but is getting bad or almost unusable with large row counts.

I have tried to investigate what is causing the twice rendering but haven't figured it out and now when I found this issue, I have strong hunch the problem is related to this issue.

**Comment by snowystinger** [comment]
Hi, yes, it's available for RAC. You can look at our own components in packages/@react-spectrum/s2 for how we do it. There are also some examples in the RAC storybook.

Documentation for Virtualization is coming soon, we recently added it to our backlog. But until it's out, looking at our source is the best way forward.

As for the double render, we render the entire collection out to a fake and fast minimal dom implementation. This collection causes a new render cycle as it is state, and that's when it gets rendered to the real dom.

**Comment by henkkasoft** [comment]
Thanks a lot for fast response.
I was able to use Virtualizer but for some reason inspecting the element tree with dev tools show that all the rows are still there?
Did I understood correctly that with Virtualizer only the visible rows should be there and that would give the performance boost when table has a lot of rows?

**Comment by snowystinger** [comment]
yes, did you set a height on your component? if you don't, the virtualizer will still render everything because it will keep having more space.

**Comment by henkkasoft** [comment]
I don't manage to get virtualizing to work. Would really appreciate the documentation or simple example. 👍

**Comment by snowystinger** [comment]
Here's the simplest example we have right now https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/stories/Table.stories.tsx#L613

**Comment by henkkasoft** [comment]
Thank you! I was able to get the Virtualizer to work and it sure remove the performance problems.

I am not totally understand how it works but for Table I used structure
```
<ResizableTableContainer style={{ maxHeight: '100vh', overflow: 'auto', scrollPaddingTop: 25 }}>
      <Virtualizer layout={layout}>
        <Table>
```

And for the ListBox
```
<Popover>
  <Virtualizer layout={layout}>
    <ListBox style={{ maxHeight: '100vh', overflow: 'auto', scrollPaddingTop: 25 }} items={items} />
```

So the virtualizing to work I needed to set maxHeight (heigth would take the space also if only few items/rows) and overflow auto, like the examples show. But I was wondering why in the Table case the maxHeight/overflow have to be set in the Virtualizer's parent component and in the ListBox it have to be set in the Virtualizer's child component?

Also the unit-tests started to fail with virtualizer. That is something I haven't figured out yet. If you have ideas also for that, I would be very grateful.

**Comment by henkkasoft** [comment]
I was able to fix table tests with this:
```
beforeEach(() => {
  vi.spyOn(window.HTMLElement.prototype, 'clientWidth', 'get').mockImplementation(() => 500);
  vi.spyOn(window.HTMLElement.prototype, 'clientHeight', 'get').mockImplementation(() => 500);
});
```

Found the idea from here:
https://github.com/adobe/react-spectrum/blob/ca0430f5150a5416fdd23398ac30ffdfa59aa766/packages/react-aria-components/test/Table.test.js

But the same did not fix the ComboBox tests. Any ideas for that?

ListBox tests use the same spyOn mocks, but that does not work with Combobox.
https://github.com/adobe/react-spectrum/blob/ca0430f5150a5416fdd23398ac30ffdfa59aa766/packages/react-aria-components/test/ListBox.test.js

And ComboBox tests don't have Virtualizer test:
https://github.com/adobe/react-spectrum/blob/ca0430f5150a5416fdd23398ac30ffdfa59aa766/packages/react-aria-components/test/ComboBox.test.js
